### PR TITLE
fix: properly mark array elements when an realm slice is updated

### PR DESCRIPTION
**Pull request by deelawn**
Addresses #1167,  #960, and #1170 

Consider the following situation:
- A slice of structs exists with a length of zero and a capacity of one
- A new struct literal is appended to the slice
- The code panics because the newly allocated struct literal was never marked as "new"

``` go
package append

import (
	"gno.land/p/demo/ufmt"
)

type T struct{ i int }

var a []T

func init() {
        a = make([]T, 0, 1)
}

func Append(i int) {
	a = append(a, T{i: i})
}

```

Invoking the `Append` function will cause a panic.

The solution is to traverse each of the array elements after slice append assignment to make sure any new or updated elements are marked as such.

This PR also includes a change to ensure that marking an object as dirty and managing references to the object are mutually exclusive.  I think this is correct but am not sure.

The changes include txtar test cases that incorporate the issue described by @tbruyelle in #1170

### CODE REVIEW

**Comment by n0izn0iz** [comment]
Do you think we can close https://github.com/gnolang/gno/pull/960 if this is merged?

**Comment by thehowl** [comment]
Great job on this!

FWIW, I'm not sure this is related to #1167 (it's different from #1170). #1167 specifically is about the underlying values of the slice turning to be "zero values", which doesn't seem to be the case here (although the fix may address that too).

I can try to check out if the behaviour in GnoChess is fixed with this PR by reverting the commit and trying to do the replication steps for that behaviour (or I can give them to you if you want to try that, I'm going a bit by memory though, don't know if they were written down).

**Comment by deelawn** [comment]
> Do you think we can close https://github.com/gnolang/gno/pull/960 if this is merged?

I think so, yes

**Comment by deelawn** [comment]
@thehowl I see your point. I didn't realize they are different issues. Though I'm not sure I completely understand the issue based on the description in #1167. Is it possible to replicate the same bad behavior using a txtar file? In any case, I'll follow up with you on this.

**Comment by thehowl** [comment]
> Is it possible to replicate the same bad behavior using a txtar file?

I've made a note, I wanted to do this already but I'll prioritize so we can find out if it's the same issue. Thanks!

**Comment by tbruyelle** [comment]
https://github.com/tbruyelle/gno/commit/71b27ee4ca6b29f735969ff3b7daf981741c3c78

So I think we can state that the bugs we have in #960, #1167 and #1170,
are all related to slice storage when its capacity is different than its
length.

@deelawn found a great way to overcome that bug, but the bug is still
there, somewhere in the VM code I think. I spent the last couple of days
trying to find it, unfortunately without success.

That said, I found a workaround, that could be also applied: when a
slice is stored, ignore any capacity different than the slice's length.

I think this is a good workaround because its one-line and because we
don't really care about storing slice with capacity higher than their
length (unless I'm missing something).

Tell me what you think.

**Comment by n0izn0iz** [comment]
that's the first workaround I attempted, but I had failing tests when I did
the capacity of a slice is not some behind-the-scene value, it can be used in "user-space" algorithms so it will alter the behavior, I'll try to write some examples here when I have more time

**Comment by deelawn** [comment]
@tbruyelle It doesn't seem like something we'd want as a long term solution, but I'm not sure why that might be.  Other than the failing tests observed by @n0izn0iz, the first thing that comes to my mind is the question of how this might affect the frequency of growing slices' underlying arrays.  But I don't know where that happens so it's just a guess 🤷 

In any case,  I believe this is close to being merged; it should address all the issues you've listed.  Or were you suggesting that there is a separate bug that should be fixed and this PR isn't addressing some underlying root cause? If that's the case then let's have a chat about the nature of the underlying bug and how we can identify where to fix it.

**Comment by tbruyelle** [comment]
@deelawn I probably spent too much time on this, I thought I had a case where your fix doesn't work but I cannot reproduce. Sorry for that.

Anyway for the memo, I think I found a way to reproduce the bug with a single transaction, using a gno file in `gnovm/tests/files/` folder and the `Realm:` instruction:

Here is the content of `gnovm/tests/files/slice_cap.gno`
```go
// PKGPATH: gno.land/r/slice_cap
package slice_cap

var a []int

func init() {
	a = make([]int, 0, 1)
}

func main() {
	a = append(a, 42)
	println(a)
}

// Output:
// slice[(42 int)]

// Realm:
// switchrealm["gno.land/r/slice_cap"]
// u[95eec26514f13645e29d560a2f0866a8dfe44dee:4]={
//     "Data": null,
//     "List": [
//         {
//             "N": "KgAAAAAAAAA=",
//             "T": {
//                 "@type": "/gno.PrimitiveType",
//                 "value": "32"
//             }
//         }
//     ],
//     "ObjectInfo": {
//         "ID": "95eec26514f13645e29d560a2f0866a8dfe44dee:4",
//         "ModTime": "4",
//         "OwnerID": "95eec26514f13645e29d560a2f0866a8dfe44dee:2",
//         "RefCount": "1"
//     }
// }
// u[95eec26514f13645e29d560a2f0866a8dfe44dee:2]={
//     "Blank": {},
//     "ObjectInfo": {
//         "ID": "95eec26514f13645e29d560a2f0866a8dfe44dee:2",
//         "IsEscaped": true,
//         "ModTime": "4",
//         "RefCount": "2"
//     },
//     "Parent": null,
//     "Source": {
//         "@type": "/gno.RefNode",
//         "BlockNode": null,
//         "Location": {
//             "File": "",
//             "Line": "0",
//             "Nonce": "0",
//             "PkgPath": "gno.land/r/slice_cap"
//         }
//     },
//     "Values": [
//         {
//             "T": {
//                 "@type": "/gno.FuncType",
//                 "Params": [],
//                 "Results": []
//             },
//             "V": {
//                 "@type": "/gno.FuncValue",
//                 "Closure": {
//                     "@type": "/gno.RefValue",
//                     "Escaped": true,
//                     "ObjectID": "95eec26514f13645e29d560a2f0866a8dfe44dee:3"
//                 },
//                 "FileName": "main.gno",
//                 "IsMethod": false,
//                 "Name": "init.0",
//                 "PkgPath": "gno.land/r/slice_cap",
//                 "Source": {
//                     "@type": "/gno.RefNode",
//                     "BlockNode": null,
//                     "Location": {
//                         "File": "main.gno",
//                         "Line": "6",
//                         "Nonce": "0",
//                         "PkgPath": "gno.land/r/slice_cap"
//                     }
//                 },
//                 "Type": {
//                     "@type": "/gno.FuncType",
//                     "Params": [],
//                     "Results": []
//                 }
//             }
//         },
//         {
//             "T": {
//                 "@type": "/gno.FuncType",
//                 "Params": [],
//                 "Results": []
//             },
//             "V": {
//                 "@type": "/gno.FuncValue",
//                 "Closure": {
//                     "@type": "/gno.RefValue",
//                     "Escaped": true,
//                     "ObjectID": "95eec26514f13645e29d560a2f0866a8dfe44dee:3"
//                 },
//                 "FileName": "main.gno",
//                 "IsMethod": false,
//                 "Name": "main",
//                 "PkgPath": "gno.land/r/slice_cap",
//                 "Source": {
//                     "@type": "/gno.RefNode",
//                     "BlockNode": null,
//                     "Location": {
//                         "File": "main.gno",
//                         "Line": "10",
//                         "Nonce": "0",
//                         "PkgPath": "gno.land/r/slice_cap"
//                     }
//                 },
//                 "Type": {
//                     "@type": "/gno.FuncType",
//                     "Params": [],
//                     "Results": []
//                 }
//             }
//         },
//         {
//             "T": {
//                 "@type": "/gno.SliceType",
//                 "Elt": {
//                     "@type": "/gno.PrimitiveType",
//                     "value": "32"
//                 },
//                 "Vrd": false
//             },
//             "V": {
//                 "@type": "/gno.SliceValue",
//                 "Base": {
//                     "@type": "/gno.RefValue",
//                     "Hash": "4ba43144bcf3cb8df3f5b60efaa9e30e1232b468",
//                     "ObjectID": "95eec26514f13645e29d560a2f0866a8dfe44dee:4"
//                 },
//                 "Length": "1",
//                 "Maxcap": "1",
//                 "Offset": "0"
//             }
//         }
//     ]
// }
```

Using the `Realm:` instruction, we can attest what is stored or not. Here there's 2 objects saved:
- `u[95eec26514f13645e29d560a2f0866a8dfe44dee:4]`  is the slice
- `u[95eec26514f13645e29d560a2f0866a8dfe44dee:2]` is the whole block/package

So this version works on this branch, while it fails on `master` because the first object `u[95eec26514f13645e29d560a2f0866a8dfe44dee:4]` is missing. This typically demonstrates that without your fix, the slice is not saved properly. Maybe that can help for better testing/refac iterations.

**Comment by deelawn** [comment]
@thehowl added a test case to #1167 that this PR does not address. I'm going to unlink that issue from this PR and will address it in a subsequent PR.  I've been discussing with @piux2 and am trying to figure out a way to rework the part of this PR that is concerned with marking appended slice elements as "new created" when the append does not result in the resizing of the underlying array.

**Comment by piux2** [comment]
This looks good to me and the fix works well. 

Two issues got fixed in this PR

 1) A updated realm slice not marked dirty:  a realm slice `foo` does not persist when `foo = append( foo, i)`, in which  `left foo` and `right foo` share the same underline array and `i` is a primitive type variable.  When the capacity expand, `left foo` and `right foo` do not share the same underline array, and it does not trigger the issue. 
 
 2) A struct variable is not marked new created:  it panics and fails to persist when a struct type variable `i` appends to a slice which they share the same array with a realm slice.  
 
 @deelawn and I looked into the options to fix the second problem at the append() in uverse.go, and breaking down to two statements in preprocess stage. We have not found better solutions than the current one, which may have more efficient alternatives. 
 
 maybe @jaekwon can take a look as well.

**Review by piux2 — APPROVED** [review]

**Review by tbruyelle — APPROVED** [review]

**Comment by jaekwon** [comment]
looking at this now. please wait for my approval to merge.

File: gnovm/pkg/gnolang/realm.go
Comment (jaekwon) [inline gnovm/pkg/gnolang/realm.go:175]: not here; this is what saveUnsavedObjectRecursively should be for...

File: gnovm/pkg/gnolang/realm.go
Comment (jaekwon) [inline gnovm/pkg/gnolang/realm.go:163]: true for rlm.MarkDirty(po), but for co, the only time we need to mark it is if after IncRefCount() the refcount is 1 (from be temporarily detached to 0). It is already marked as deleted, but it won't be deleted when the time comes if refcount > 0. When refcount becomes 1 its ownerID probably changed. If the refcount ends up being > 1, it will be saved as escaped.

**Comment by jaekwon** [comment]
Maybe append() in uverse isn't property setting the array as dirty by calling DidUpdate.

NOTE: The design consideration for DidUpdate is that it must be fast. 
The idea is to just keep enough information (by appending to newCreated/newEscaped/newDeleted/updated etc) to be processed in batch later at the end of the transaction. So we shouldn't be iterating over arrays here, since it might not even be useful if say the array is dropped later.

**Comment by jaekwon** [comment]
Yeah the above patch seems to fix the problem for test 1167.

**Comment by tbruyelle** [comment]
I tested @jaekwon 's fix, so without any other changes from this branch, directly on master. 

The [above challenge](https://github.com/gnolang/gno/pull/1305#issuecomment-1789805642) I wrote a few days ago now works as expected. The expected `// Realm:` is exactly the same.

But the append_bug.sh script that I wrote [in this PR](https://github.com/gnolang/gno/pull/1170) fails on the ultimate `Render` call, with the following error: 

```
+ gnokey maketx call -pkgpath gno.land/r/demo/bug/append -func Render -gas-fee 1000000ugnot -gas-wanted 2000000 -send '' -broadcast -chainid dev -args '' -remote 127.0.0.1:26657 test1
Enter password.
--= Error =--
Data: unexpected object with id 7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:9
Msg Traces:
    0  /home/tom/src/gno/tm2/pkg/crypto/keys/client/addpkg.go:214 - deliver transaction failed: log:msg:0,success:false,log:--= Error =--
Data: &errors.errorString{s:"unexpected object with id 7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:9"}
Msg Traces:
    0  /home/tom/src/gno/gno.land/pkg/sdk/vm/keeper.go:266 - VM call panic: unexpected object with id 7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:9
Machine:
    CheckTypes: false
	Op: [OpHalt OpBody OpForLoop OpAddAssign OpPrecall OpSelector]
	Values: (len: 4)
          #3 (slice[(struct{(2 int)} gno.land/r/demo/bug/append.T),(ref(7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:9) gno.land/r/demo/bug/append.T),(ref(7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:10) gno.land/r/demo/bug/append.T)] []gno.land/r/demo/bug/append.T)
          #2 ("%!!(MISSING)d(MISSING)</br>" string)
          #1 (Sprintf func(format string,args ...interface{})( string))
          #0 (Render func(.arg_0 string)( string))
	Exprs:
          #1 a<VPBlock(4,4)>[i<VPBlock(1,0)>].i
          #0 ufmt<VPBlock(3,0)>.Sprintf((const ("%!!(MISSING)d(MISSING)</br>" string)), a<VPBlock(4,4)>[i<VPBlock(1,0)>].i)
	Stmts:
          #1 bodyStmt[0/0/1]=(end)
          #0 bodyStmt[0/0/2]=return s<VPBlock(1,2)>*
	Blocks:
          @(1) Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc00142cd20,Source:for i<VPBlock(1,0)> := (const (0...,Parent:0xc00142cb40)
            i: (1 int)
 (s vals) @(1) Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc004db4c20,Source:for i<VPBlock(1,0)> := (const (0...,Parent:0xc00553a420)
            i: (0 int)
 (s typs) @(1) [int]
          @(2) Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc00142cb40,Source:func Render(.arg_0 (const-type s...,Parent:0xc00142c1e0)
            .arg_0: ("" string)
            .res_0: ( string)
            s: ("2</br>" string)
 (s vals) @(2) Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc00553a420,Source:func Render(.arg_0 (const-type s...,Parent:0xc006016860)
            .arg_0: ( string)
            .res_0: ( string)
            s: (undefined)
 (s typs) @(2) [string string string]
          @(3) Block(ID:7bc001193bf94ce61dc11a3b5f7b741a56e22e9d:3,Addr:0xc00142c1e0,Source:ref(gno.land/r/demo/bug/append/a...,Parent:0xc00142c000)
            (RefNode names not shown)
 (s vals) @(3) Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc006016860,Source:file{ package append; import ufm...,Parent:0xc006016b20)
            ufmt: (package(ufmt gno.land/p/demo/ufmt) package{})
 (s typs) @(3) [package{}]
          @(4) gno.land/r/demo/bug/append
	Blocks (other):
          #1 Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc00142cb40,Source:func Render(.arg_0 (const-type s...,Parent:0xc00142c1e0)
            .arg_0: ("" string)
            .res_0: ( string)
            s: ("2</br>" string)
 (static) #1 Block(ID:0000000000000000000000000000000000000000:0,Addr:0xc00553a420,Source:func Render(.arg_0 (const-type s...,Parent:0xc006016860)
            .arg_0: ( string)
            .res_0: ( string)
            s: (undefined)
	Frames:
          #1 [FRAME LABEL:  2/1/0/1/2]
          #0 [FRAME FUNC:Render RECV:(undefined) (1 args) 1/0/0/0/1 LASTPKG:main LASTRLM:Realm(nil)]
	Realm:
	  gno.land/r/demo/bug/append
	Exceptions:
	  []
	  

Stack Trace:
    0  /home/tom/src/gno/tm2/pkg/errors/errors.go:20
    1  /home/tom/src/gno/gno.land/pkg/sdk/vm/keeper.go:266
    2  /usr/lib/go/src/runtime/panic.go:914
    3  /home/tom/src/gno/gnovm/pkg/gnolang/store.go:246
    4  /home/tom/src/gno/gnovm/pkg/gnolang/values.go:2479
    5  /home/tom/src/gno/gnovm/pkg/gnolang/values.go:356
    6  /home/tom/src/gno/gnovm/pkg/gnolang/values.go:438
    7  /home/tom/src/gno/gnovm/pkg/gnolang/values.go:1906
    8  /home/tom/src/gno/gnovm/pkg/gnolang/op_expressions.go:33
    9  /home/tom/src/gno/gnovm/pkg/gnolang/machine.go:1179
   10  /home/tom/src/gno/gnovm/pkg/gnolang/machine.go:636
   11  /home/tom/src/gno/gno.land/pkg/sdk/vm/keeper.go:272
   12  /home/tom/src/gno/gno.land/pkg/sdk/vm/handler.go:64
   13  /home/tom/src/gno/gno.land/pkg/sdk/vm/handler.go:29
   14  /home/tom/src/gno/tm2/pkg/sdk/baseapp.go:644
   15  /home/tom/src/gno/tm2/pkg/sdk/baseapp.go:823
   16  /home/tom/src/gno/tm2/pkg/sdk/baseapp.go:580
   17  /home/tom/src/gno/tm2/pkg/bft/abci/client/local_client.go:82
   18  /home/tom/src/gno/tm2/pkg/bft/proxy/app_conn.go:73
   19  /home/tom/src/gno/tm2/pkg/bft/state/execution.go:253
   20  /home/tom/src/gno/tm2/pkg/bft/state/execution.go:102
   21  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1347
   22  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1275
   23  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1221
   24  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1252
   25  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1637
   26  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:1483
   27  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:691
   28  /home/tom/src/gno/tm2/pkg/bft/consensus/state.go:650
   29  /usr/lib/go/src/runtime/asm_amd64.s:1650
--= /Error =--
,events:[]
--= /Error =--
```

**Comment by deelawn** [comment]
@tbruyelle I'm seeing the same and am currently debugging this.  The issue seems to lie with the newly escaped object not being persisted. So when the final append happens it overwrites the same reference value at array element three. Unfortunately this results in the deletion of the object so the reference that exists in the second element of the array refers to an object that no longer exists, hence the error when it tries to load it into memory when Render is called.

**Comment by thehowl** [comment]
Yeah agreed that we don't really want to do a for loop on `DidUpdate` if we can somehow avoid it.

I tried applying your patch @jaekwon but, as with thomas, the txtar test fails with the same error as before.

I'd dig deeper but I know that would entail probably a day-long adventure, so I'll leave it to @deelawn to try to figure out a better solution, trying to keep DidUpdate loopless :)

**Comment by deelawn** [comment]
My most recent commit uses Jae's suggested changes along with one small addition of my own to ensure that `RefValue` type elements have their underlying value copied rather than the reference.  This was causing the `append` test case that has been mentioned in this PR to fail.  With this commit, both the append test case and the test case from #1167 are passing.  I'll need to do some more testing with perceived edge/problem cases and review the append code to see what else may need to be added for some of the other append codepaths.  

@jaekwon , can you have a look and see if the `DeepCopy` method solution for `RefValue` types is an acceptable solution in this case?

File: gnovm/pkg/gnolang/realm.go
Comment (deelawn) [reply gnovm/pkg/gnolang/realm.go]: reverted

File: gnovm/pkg/gnolang/realm.go
Comment (deelawn) [reply gnovm/pkg/gnolang/realm.go]: reverted

**Comment by jaekwon** [comment]
looking

File: gnovm/pkg/gnolang/values.go
Comment (jaekwon) [inline gnovm/pkg/gnolang/values.go:1019]: can't we do store.GetObject(cv.ObjectID) directly here?

File: gnovm/pkg/gnolang/values.go
Comment (jaekwon) [inline gnovm/pkg/gnolang/values.go:1013]: This is a misnomer. It doesn't do a deep copy, a deep copy should deep copy all children recursively. It only does a slightly deeper copy if tv happens to be a RefValue. But this function is useful at least in the context of uverse, so I would call this "unrefCopy" and not even expose it.

Comment (jaekwon) [reply gnovm/pkg/gnolang/values.go:1013]: unrefCopy makes *a* copy of the underlying value, but first unreferences if RefValue.

**Comment by jaekwon** [comment]
What about this part?

// XXX find other uses of native copy()
// and replace with .Copy(alloc) as necessary.

Now we have to look at the codebase and see if there are in the very least other uverse functions that need the same fixing... they all seem to use Assign2() which calls DidUpdate already, which is good. 

Any other places besides uverse? I don't think so.

**Review by jaekwon — APPROVED** [review]
pre-approving assuming all comments addressed or please re-assign to me

File: gnovm/pkg/gnolang/uverse.go
Comment (piux2) [inline gnovm/pkg/gnolang/uverse.go:324]: Do we need to call DidUpdate here? Is there a testing case for this?

File: gnovm/pkg/gnolang/uverse.go
Comment (piux2) [inline gnovm/pkg/gnolang/uverse.go:315]: Do we need to call DidUpdate here? Is there a testing case for this?

**Pull request merged**
